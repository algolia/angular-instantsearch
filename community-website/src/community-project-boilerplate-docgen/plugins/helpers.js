const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
