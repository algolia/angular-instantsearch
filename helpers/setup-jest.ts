import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
