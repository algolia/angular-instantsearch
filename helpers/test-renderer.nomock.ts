import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
