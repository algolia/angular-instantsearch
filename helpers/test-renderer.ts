import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
