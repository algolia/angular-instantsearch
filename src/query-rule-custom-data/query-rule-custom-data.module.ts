import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisQueryRuleCustomData } from './query-rule-custom-data';

@NgModule({
  declarations: [NgAisQueryRuleCustomData],
  entryComponents: [NgAisQueryRuleCustomData],
  exports: [NgAisQueryRuleCustomData],
  imports: [CommonModule],
})
export class NgAisQueryRuleCustomDataModule {}
