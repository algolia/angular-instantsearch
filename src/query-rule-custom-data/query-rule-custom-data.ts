import {
  Component,
  ContentChild,
  TemplateRef,
  Inject,
  forwardRef,
  Input,
} from '@angular/core';

import { connectQueryRules } from 'instantsearch.js/es/connectors';

import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';

@Component({
  selector: 'ais-query-rule-custom-data',
  template: `
    <div [class]="cx()">
      <ng-container *ngTemplateOutlet="template; context: templateContext">
      </ng-container>

      <div *ngIf="!template">
        <div *ngFor="let item of state.items">
          <pre>{{ item | json }}</pre>
        </div>
      </div>
    </div>
  `,
})
export class NgAisQueryRuleCustomData extends BaseWidget {
  @ContentChild(TemplateRef) public template: any;

  @Input() public transformItems?: (items: any[]) => any[];

  public state = {
    items: [],
  };

  get templateContext() {
    return {
      items: this.state.items,
    };
  }

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super('QueryRuleCustomData');
  }

  public ngOnInit() {
    this.createWidget(connectQueryRules, {
      transformItems: this.transformItems,
    });

    super.ngOnInit();
  }
}
