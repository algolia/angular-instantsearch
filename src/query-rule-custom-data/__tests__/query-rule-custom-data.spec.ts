import { createRenderer } from '../../../helpers/test-renderer';
import { NgAisQueryRuleCustomData } from '../query-rule-custom-data';

describe('QueryRuleCustomData', () => {
  let render;
  let createWidget;

  beforeEach(() => {
    jest.clearAllMocks();
    createWidget = jest.spyOn(
      NgAisQueryRuleCustomData.prototype,
      'createWidget'
    );

    render = (template: string, state?: object) =>
      createRenderer({
        defaultState: state,
        template,
        TestedWidget: NgAisQueryRuleCustomData,
      })({});
  });

  it('should render without state', () => {
    const template = `<ais-query-rule-custom-data></ais-query-rule-custom-data>`;
    const state = undefined;

    const fixture = render(template, state);

    expect(fixture).toMatchSnapshot();
  });

  it('should render with empty state', () => {
    const template = `<ais-query-rule-custom-data></ais-query-rule-custom-data>`;
    const state = {};

    const fixture = render(template, state);

    expect(fixture).toMatchSnapshot();
  });

  it('should render with state', () => {
    const template = `<ais-query-rule-custom-data></ais-query-rule-custom-data>`;
    const state = { items: [{ banner: '1.jpg' }, { banner: '2.jpg' }] };

    const fixture = render(template, state);

    expect(fixture).toMatchSnapshot();
  });

  it('should render with custom templates', () => {
    const template = `
<ais-query-rule-custom-data>
  <ng-template let-items="items">
    <div *ngFor="let item of items">
      <img src="{{item.banner}}" alt="{{item.title}}">
    </div>
  </ng-template>
</ais-query-rule-custom-data>
    `;
    const state = {
      items: [
        { title: 'Banner 1', banner: '1.jpg' },
        { title: 'Banner 1', banner: '2.jpg' },
      ],
    };

    const fixture = render(template, state);

    expect(fixture).toMatchSnapshot();
  });

  it('passes transformItems', () => {
    const transformItems = 'transformItems';

    const fixture = render(
      `<ais-query-rule-custom-data
          [transformItems]="'${transformItems}'"
        >
        </ais-query-rule-custom-data>`,
      { items: [{ banner: '1.jpg' }, { banner: '2.jpg' }] }
    );

    expect(createWidget.mock.calls[0][1].transformItems).toEqual(
      transformItems
    );
  });
});
