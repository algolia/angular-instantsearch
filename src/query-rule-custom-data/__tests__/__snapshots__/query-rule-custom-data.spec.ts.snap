// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryRuleCustomData should render with custom templates 1`] = `
<ng-component
  testedWidget={[Function NgAisQueryRuleCustomData]}
>
  <ais-instantsearch>
    <ais-query-rule-custom-data>
      <div
        class="ais-QueryRuleCustomData"
      >
        
        
        <div>
          <img
            alt="Banner 1"
            src="1.jpg"
          />
        </div>
        <div>
          <img
            alt="Banner 1"
            src="2.jpg"
          />
        </div>
        
        
      </div>
    </ais-query-rule-custom-data>
  </ais-instantsearch>
</ng-component>
`;

exports[`QueryRuleCustomData should render with empty state 1`] = `
<ng-component
  testedWidget={[Function NgAisQueryRuleCustomData]}
>
  <ais-instantsearch>
    <ais-query-rule-custom-data>
      <div
        class="ais-QueryRuleCustomData"
      >
        
        
        <div>
          
        </div>
      </div>
    </ais-query-rule-custom-data>
  </ais-instantsearch>
</ng-component>
`;

exports[`QueryRuleCustomData should render with state 1`] = `
<ng-component
  testedWidget={[Function NgAisQueryRuleCustomData]}
>
  <ais-instantsearch>
    <ais-query-rule-custom-data>
      <div
        class="ais-QueryRuleCustomData"
      >
        
        
        <div>
          
          <div>
            <pre>
              {
    "banner": "1.jpg"
  }
            </pre>
          </div>
          <div>
            <pre>
              {
    "banner": "2.jpg"
  }
            </pre>
          </div>
        </div>
      </div>
    </ais-query-rule-custom-data>
  </ais-instantsearch>
</ng-component>
`;

exports[`QueryRuleCustomData should render without state 1`] = `
<ng-component
  testedWidget={[Function NgAisQueryRuleCustomData]}
>
  <ais-instantsearch>
    <ais-query-rule-custom-data>
      <div
        class="ais-QueryRuleCustomData"
      >
        
        
        <div>
          
        </div>
      </div>
    </ais-query-rule-custom-data>
  </ais-instantsearch>
</ng-component>
`;
